### PR TITLE
A few more improvements to the strict positivity checker

### PR DESCRIPTION
**Pull request by nikswamy**
1. A bug fix

The strictly_positive attribute in binders was not properly handled in subtyping, meaning that you could subsume a `Type -> Type` to `([@@@strictly_positive]a:Type) -> Type`, effectively claiming that a function was strictly positive in its argument when it was not. This is now fixed. Binders now have an additional field, binder_positivity, which records any positivity-related attributes, and subtyping allows a binder_positivity qual to be silently dropped, but not introduced.

Note, as far as the reflection API goes, a binder_view still only has 3 fields, sort, qual, and attributes. When moving to and from the reflection binder_view and the internal representation of binders, the binder_positivity field is populated or reflected in the attributes. See Syntax.Util.parse/encode_postivity_attributes

2.  Some improvements to the positivity checker, which now also handles recursive definitions and match

E.g.,

```
let rec ffff ([@@@strictly_positive] a:Type) (x:ℕ) = if x = 0 then a else unit → ffff a (x - 1)
```

3. We have a new attribute [@@@unused] which is stricter that strictly_positive. A binder is unused in a definition if
* it does not appear in the definition
* or, it appears only as an argument to a function whose corresponding binder is marked with the unused attribute

4. Several types in the library now have strictly_positive or unused attributes, including
* Strictly_positive: erased, seq, map, dependent_map, partial_map, vector
* unused: Steel.Reference.ref, Steel.ST.Reference.ref etc.

### CODE REVIEW

File: src/syntax/FStar.Syntax.Util.fst [hunk 1]
```diff
@@ -2497,3 +2497,53 @@ let flatten_refinement t =
     | _ -> t
   in
   aux t false
+
+let contains_strictly_positive_attribute (attrs:list attribute)
```
Comment (aseemr) [inline src/syntax/FStar.Syntax.Util.fst:2501]: Minor nit: Syntax.Util already has `has_attribute` which could be used here and in `contains _unused_attribute`.

File: src/typechecker/FStar.TypeChecker.Core.fst [hunk 2]
```diff
@@ -1018,6 +1025,7 @@ let rec check_relation (g:env) (rel:relation) (t0 t1:typ)
       | Tm_abs([b0], body0, _), Tm_abs([b1], body1, _) ->
         check_relation g EQUALITY b0.binder_bv.sort b1.binder_bv.sort;!
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.Core.fst:1026]: Not related to this PR, but surprised that for lambdas, we are checking binder bv sorts equality, irrespective of the input relation; whereas for arrows we check relation (with contravariance).

**Comment by aseemr** [comment]
Looks good to me, thanks!

File: src/syntax/FStar.Syntax.Util.fst (hunk 1 above)
Comment (nikswamy) [reply src/syntax/FStar.Syntax.Util.fst:2504]: fixed

**Pull request merged**